### ISSUE TITLE
[designspaceLib] Deduces base master wrong if axis element is missing

### ISSUE THREAD

**Issue by behdad** [fonttools/fonttools#1228]
Should fall back to checking for copyinfo stuff, like varLib does.

I think I like to replace most of the logic in designspaceLib with the one from varLib.

**Comment by behdad** [comment]
So. We need an answer to "who owns / maintains fontTools.designspaceLib?" @LettError @belluzj

**Comment by LettError** [comment]
Anything specific that prompts this question, @behdad?

**Comment by behdad** [comment]
> Anything specific that prompts this question, @behdad?

Well. This bug and https://github.com/fonttools/fonttools/issues/1226

For most of fontTools modules, we have individuals who own it and are responsible to unblock others. Would be nice to have the same in this module.

**Comment by behdad** [comment]
I understand. I was primarily asking if Jany can own it. If neither of you have time, I will.

**Comment by belluzj** [comment]
I'm sorry but I'm moving on to other tasks now, so I won't have much time for improving designspaceLib (also, I wouldn't want to "own" it anyway, I think Erik was supposed to stay the "benevolent dictator" as per https://github.com/LettError/designSpaceDocument/issues/28#issue-274587260)

**Comment by LettError** [comment]
I thought integrating designspacedocument into fonttools would make it easier to take PRs from others. Responding to many of the recent issues concerning the Noto workflow, integrating with GlyphsLib etc. are quite outside my field. I'm concerned for big sweeping changes or fixes for Noto specific issues that will break other tools that rely on designspaces, but barring those I don't want to block anything.

**Comment by LettError** [comment]
So.. about finding the base master / default master / neutral master. 

[checkDefault()](https://github.com/fonttools/fonttools/blob/master/Lib/fontTools/designspaceLib/__init__.py#L1225) has issues. Earlier designspaces did not have axes defined and that required some guessing. I think by now we can safely assume the axes are defined and we can remove the axis reconstruction. We also don't have to compare it to whatever mutatormath was doing.

So what if `checkDefault` just did the following:

1. Find a master that is on the _default value of all the axes._
2. Find a master with the `copyInfo` flag.
3. Fail and complain.

**Comment by anthrotype** [comment]
>  think by now we can safely assume the axes are defined and we can remove the axis reconstruction

when axes are missing, varLib doesn't attempt to reconstruct them, but simply falls back to a hard-coded standard_axis_map (which only makes sense for OT font variations).

Wouldn't it be OK by now to simply just require that the axes element be there, and use that to find the the base master (point 1.)? I would just retire the copyInfo method altogether.

**Comment by LettError** [comment]
> Wouldn't it be OK by now to simply just require that the axes element be there, and use that to find the the base master (point 1.)? I would just retire the copyInfo method altogether.

I agree. @readroberts or @miguelsousa what do you think?

**Comment by miguelsousa** [comment]
SGTM.

**Comment by readroberts** [comment]
I agree also.

**Comment by anthrotype** [comment]
> we can remove the axis reconstruction

the `DesignSpaceDocument.checkAxes` method (called by `check` method) currently is used to "repair" the designspace, by adding axes that are used in the sources/instances but are missing from the `axes` element.
Should that be dropped as well?

So basically those checkAxes and checkDefault methods should now raise if axes element is missing, or if there is any axis referenced in sources or instances or rules etc. which is not also defined in the axes top element, or if there is no default master?

I think that alone warrants a version bump for the designspace document format.

Another option could be issue some deprecation warnings first, then drop support for designspaces-without-explicit-axes a few versions later.

**Comment by LettError** [comment]
@anthrotype deprecation warnings are probably cautious and necessary steps.

**Comment by anthrotype** [comment]
> deprecation warnings are probably cautious and necessary steps

I'm not sure what we are proposing to keep or deprecate here.
We said we want to require that the axes element be defined; in which case there is only one base master: i.e. the one with default locations on all the axes (0 in normalized coordinates).
Under these circumstances, the copyInfo flag becomes redundant for the sake of defining the neutral (though it may still be used for ufo editors to define from where to copy the fontinfo.plist in the generated ufo instances).

**Comment by anthrotype** [comment]
in fonttools 3.27, we

> Removed checkDefault, checkAxes methods, and any kind of guessing about the axes when the <axes> element is missing. The default master is expected at the intersection of all default
values for each axis (#1254, #1255, #1267)

are we all happy and can we now close this or is there anything else to be done here?

**Comment by behdad** [comment]
So, axes is now mandatory? Sounds good to me.
